pkg/hub/addon/discovery_controller.go

```diff
@@ -0,0 +1,213 @@
+package addon
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	addonv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	addoninformerv1alpha1 "open-cluster-management.io/api/client/addon/informers/externalversions/addon/v1alpha1"
+	addonlisterv1alpha1 "open-cluster-management.io/api/client/addon/listers/addon/v1alpha1"
+	clientset "open-cluster-management.io/api/client/cluster/clientset/versioned"
+	clusterv1informer "open-cluster-management.io/api/client/cluster/informers/externalversions/cluster/v1"
+	clusterv1listers "open-cluster-management.io/api/client/cluster/listers/cluster/v1"
+)
+
+const (
+	addOnFeaturePrefix = "feature.open-cluster-management.io/addon-"
+)
+
+// addOnFeatureDiscoveryController monitors ManagedCluster and its ManagedClusterAddOns on hub and
+// create/update/delete labels of the ManagedCluster to reflect the status of addons.
+type addOnFeatureDiscoveryController struct {
+	clusterClient clientset.Interface
+	clusterLister clusterv1listers.ManagedClusterLister
+	addOnLister   addonlisterv1alpha1.ManagedClusterAddOnLister
+	recorder      events.Recorder
+}
+
+// NewAddOnFeatureDiscoveryController returns an instance of addOnFeatureDiscoveryController
+func NewAddOnFeatureDiscoveryController(
+	clusterClient clientset.Interface,
+	clusterInformer clusterv1informer.ManagedClusterInformer,
+	addOnInformers addoninformerv1alpha1.ManagedClusterAddOnInformer,
+	recorder events.Recorder,
+) factory.Controller {
+	c := &addOnFeatureDiscoveryController{
+		clusterClient: clusterClient,
+		clusterLister: clusterInformer.Lister(),
+		addOnLister:   addOnInformers.Lister(),
+		recorder:      recorder,
+	}
+
+	return factory.New().
+		WithInformers(clusterInformer.Informer()).
+		WithInformersQueueKeyFunc(
+			func(obj runtime.Object) string {
+				accessor, _ := meta.Accessor(obj)
+				return accessor.GetName()
+			},
+			clusterInformer.Informer()).
+		WithInformersQueueKeyFunc(
+			func(obj runtime.Object) string {
+				key, _ := cache.MetaNamespaceKeyFunc(obj)
+				return key
+			},
+			addOnInformers.Informer()).
+		WithSync(c.sync).
+		ToController("AddOnFeatureDiscoveryController", recorder)
+}
+
+func (c *addOnFeatureDiscoveryController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	queueKey := syncCtx.QueueKey()
+	namespace, name, err := cache.SplitMetaNamespaceKey(queueKey)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return nil
+	}
+
+	if len(namespace) > 0 {
```

Comment:
it needs doc here, otherwise it is really strange

Reply:
Done.